### PR TITLE
FIREFLY-1694,1105,1784: Improve unit recognition of SpectrumDM

### PR DESCRIPTION
**Pull request by jaladh-singhal**
Fixes [FIREFLY-1694](https://jira.ipac.caltech.edu/browse/FIREFLY-1694), [FIREFLY-1105](https://jira.ipac.caltech.edu/browse/FIREFLY-1105), [FIREFLY-1784](https://jira.ipac.caltech.edu/browse/FIREFLY-1784)

Following task-list is derived from the text and comments in above tickets (+ conversations about this work in meetings):

- [x] Make firefly recognize flux units in `erg/s/cm^2` and convert to equivalent units: `W/m^2/Hz <-> Jy <-> erg/s/cm^2/Hz`. `erg/s/cm^2/A` can't be converted to flux density in frequency space so added one more wavelength space unit `W/m^2/um`.
- [x] Show units in options dialog even when firefly can't recognize them: instead of dropdown of convertible units, this will be a readonly ~field with a tip~ text.
- [x] Make firefly recognize alias of same unit: `Angstrom = angstrom = A`, `erg/s/cm^2/Angstrom = erg/s/cm2/Angstrom = erg/s/cm**2/Angstrom = erg.s**-1.cm**-2.Angstrom**-1`, etc.
- [x] Make firefly display the recognized units on chart axes in shorter latex syntax i.e., if it's recognized by firefly, it will match the dropdown label + other label addition that firefly makes like redshift correction
- [x] Make firefly recognize units prefixed with a scalar factor, also convert them so it will be a new entry in dropdown
- [x] Fix bug in axis label doesn't update with correct units when spectrum is single trace
- [x] Change disabled input fields to just simple text
- [x] Add lamda, F_lambda symbols that appear in chart to dialog too indicate they map to column name
- [x] Add unit tests
- [x] wording changes in spectra file upload

## Testing
https://fireflydev.ipac.caltech.edu/firefly-1694-1105/firefly/

Upload test files (attached in the 1694 ticket), first on irsaviewer ops and then on this build. Open chart dialogs for each file and notice how they differ between irsaviewer and this build. Change unit dropdowns (and redshift correction ones), it should update the column field with correct formula and chart axes label with correct label (once you hit apply)

### CODE REVIEW

**Comment by jaladh-singhal** [comment]
For tracking everything at one place:

Comments from @loitly on Jun 17, 2025:

> However, since I’ve already taken a quick look and likely won’t do a full review—here are my thoughts:
>  
> - The improvements to the conversion logic make it more robust—nice work!
>  
> - I like the added documentation.  But, one thing I didn’t see (or may have missed) is a top-level explanation of how the main data structures relate to each other, e.g., UnitMetadata, UnitXref, and Measurement.
>  
> - A unit test for this file would also be helpful.  Something like sprintf-test.js that shows the expected input and output for each function.  (not all, just the important ones)
>  
> I haven’t looked at it in detail.  These are just my initial impressions, so take them with a grain of salt.

**Comment by lrebull** [comment]
minor word changes, not directly related to this issue so i made a new ticket: https://jira.ipac.caltech.edu/browse/FIREFLY-1784 Jaladh, i added you as a watcher since trey is out and i don't know if this is an easy fix or not.

This looks really nice! i have also asked vandana to take a look.

**Comment by vandesai1** [comment]
Testing at https://fireflydev.ipac.caltech.edu/firefly-1694-1105/firefly/

Test 1. Open a Euclid combspec file as delivered. 

a. The 1D spectrum extensions are correctly plotted as spectra.

**b. In the chart options, "Trace Style" is now at the top, and no longer in the Trace options. That's kinda strange.**

c. The conversion between erg/s/cm^2/Angstrom and W/m^2/micron looks correct.


Test 2. I uploaded a file that I modified (modified_euclid_spectrum.vot). The flux has units of 1e-16erg.s**-1.cm**-2.Angstrom**-1

a. This was recognized as a spectrum and plotted correctly.

b. The unit conversion looks correct, even with the factor out front! Nice!

Test 3: I uploaded another modified file (mod2_euclid_spectrum.vot), in which I replaced all instances of "Angstrom" to "Ang".

a. Firefly still recognized this as a spectrum. good.

b. The units were displayed in the chart area, but there weren't any conversion options. On ops, the units were not shown. good.

Test 4. I uploaded mod3.vot, which has units of 1e-16erg.s**-1.cm^-2.Angstrom**-1

**Firefly did NOT understand this.** 

Test 5. I uploaded mod4.vot, which has units of 1e-16erg.s**-1.cm-2.Angstrom**-1

**Firefly did NOT understand this.**

Test 5.  I uploaded mod5.vot, which has units of 1e-16erg/s/cm2/A

Worked fine!

Test 6.  I uploaded mod6.vot, which has units of 1e-16erg/s/cm**2/A

Worked fine!

Test 7.  I uploaded mod7.vot, which has units of 1e-16erg/s/cm^2/A

Worked fine!

**Comment by jaladh-singhal** [comment]
@vandesai1 thanks for the detailed testing. It seems to me you’re mostly happy with your test results except the bolded ones:
> In the chart options, “Trace Style” is now at the top, and no longer in the Trace options. That’s kinda strange.

1. It was never inside in Trace options in ops too (not sure why but that's the case for scatter chart type too: go to "overplot new trace"). And about moving it to top, I did it because we have section separators for each X and Y axis now so if I kept it below, it would seem it's part of Y axis section but it's not.

ops | this PR build
--- | ---
image | image

I also feel it should be inside Trace options collapsible, I can move it there unless we had an historical reason to keep it outside?

---
> Test 4. I uploaded mod3.vot, which has units of 1e-16erg.s**-1.cm^-2.Angstrom**-1
> Test 5. I uploaded mod4.vot, which has units of 1e-16erg.s**-1.cm-2.Angstrom**-1
> Firefly did NOT understand this.


2. Yes firefly can't recognize mixed notations of powers (`u**2`, `u^2`, `u2`) and in this VO dot product representation ([VOUnits Grammar](https://ivoa.net/documents/VOUnits/20231215/REC-VOUnits-1.1.pdf) on page 42) it strictly expects `**` for powers. Do you want Firefly to recognize your test units containing mixed notations?

**Comment by lrebull** [comment]
I can see your code changes (oooh i'm getting more confident in understanding how to use github!)
"Some table appears to be a spectrum" sounds very strange too. 
How about "Some tables appear to be spectra."
(the other change looks good!)

**Comment by jaladh-singhal** [comment]
> I can see your code changes (oooh i'm getting more confident in understanding how to use github!)

Yay! 

> "Some table appears to be a spectrum" sounds very strange too.
How about "Some tables appear to be spectra."

Done!

**Comment by vandesai1** [comment]
> 2. Yes firefly can't recognize mixed notations of powers (u**2, u^2, u2) and in this VO dot product representation ([VOUnits Grammar](https://ivoa.net/documents/VOUnits/20231215/REC-VOUnits-1.1.pdf) on page 42) it strictly expects ** for powers. Do you want Firefly to recognize your test units containing mixed notations?

Nah, I think it's ok like this.

I need to eat something and then look at #1.

**Comment by vandesai1** [comment]
>1.  It was never inside in Trace options in ops too (not sure why but that's the case for scatter chart type too: go to "overplot new trace"). And about moving it to top, I did it because we have section separators for each X and Y axis now so if I kept it below, it would seem it's part of Y axis section but it's not.

I agree that it's strange that the trace style has appeared in the Y axis section. I can't remember why that was there. Maybe we shouldn't change it without a ccb, since it wasn't part of any of these tickets?

File: src/firefly/js/charts/__tests__/SpectrumUnitConversion-test.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+    test('canUnitConv', () => {
+        expect(canUnitConv({from: 'Hz', to: 'A'})).toBe(false);
+        expect(canUnitConv({from: 'Hz', to: 'GHz'})).toBe(true);
+    });
+
+    // TODO: write tests for all exported functions?
+    test('getUnitConvExpr', () => {
+
+    });
```
Comment (loitly) [inline src/firefly/js/charts/__tests__/SpectrumUnitConversion-test.js]: What I’m looking for here is documentation that demonstrates how this module is intended to be used and what results to expect(documenting behavior).
Some usage examples would help so I don’t have to read through the code. Simple examples are great, but a few trickier ones would be even better.

**Review by loitly — APPROVED** [review]
Code looks good. About the Trace Style, I think it was mainly to keep frequently changed settings outside the collapsible panel. Whatever we decide, I think we should do the same for the Scatter plot options panel too.

File: src/firefly/js/charts/__tests__/SpectrumUnitConversion-test.js (hunk 1 above)
Comment (jaladh-singhal) [reply src/firefly/js/charts/__tests__/SpectrumUnitConversion-test.js]: Done! It actually helped me find few corner cases that my code was overlooking.

**Comment by jaladh-singhal** [comment]
Since CI will run after this PR is merged, posting my local output:
```
yarn run v1.22.22
$ node ../../node_modules/jest/bin/jest --passWithNoTests SpectrumUnitConversion
 PASS  js/charts/__tests__/SpectrumUnitConversion-test.js
  Test SpectrumUnitConversion
    ✓ canUnitConv (5ms)
    ✓ getUnitConvExpr (6ms)
    ✓ getUnitInfo: LAMBDA (2ms)
    ✓ getUnitInfo: F_LAMBDA (2ms)
    ✓ getUnitInfo: other measurements (1ms)
    ✓ getXLabel
    ✓ getMeasurementLabel (1ms)

Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        7.143s
Ran all test suites matching /SpectrumUnitConversion/i.
✨  Done in 7.80s.
```
